### PR TITLE
reimbursableExpenses Ui implementation

### PR DESCRIPTION
**Pull request by DhaaraniCIT**
Implementation Working video
https://github.com/fylein/fyle-integrations-settings-app/assets/55541808/a4a06f23-170d-4211-9a00-581f786dea39

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
Screenshot 2023-10-19 at 5 58 19 PM
Screenshot 2023-10-19 at 5 58 27 PM

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html [hunk 1]
```diff
@@ -1 +1,244 @@
...
+                            [isFieldMandatory]="true"
+                            [mandatoryErrorListName]="'export module'"
+                            [label]="'How should the expenses be exported?'"
+                            [options]="expensesExportTypes"
+                            [iconPath]="'expense'"
+                            [placeholder]="'Select expense export module'"
+                            [formControllerName]="'reimbursableExportType'"
+                            [appName]="appName">
+                        </app-configuration-select-field>
+                        <div *ngIf="exportSettingForm.value.reimbursableExportType===expensesExportTypes[1].value">
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html:38]: let's hardcode value, expensesExportTypes[1].value is not readable. Use enum for comparison

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html [hunk 2]
```diff
@@ -1 +1,244 @@
...
+                            <app-configuration-select-field
+                                [form]="exportSettingForm"
+                                [isFieldMandatory]="true"
+                                [mandatoryErrorListName]="'Default Credit Account Name'"
+                                [label]="'Set the Default Credit Account as?'"
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html:43]: I don't think this we need it in reimburse section, will check once API folks confirms

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html [hunk 3]
```diff
@@ -1 +1,244 @@
...
+                            [isFieldMandatory]="true"
+                            [mandatoryErrorListName]="'employee field'"
+                            [label]="'How are your Employees represented in Sage 300 CRE?'"
+                            [subLabel]="'Select how you represent your employees in Sage Intacct. This would help to export the expenses to the correct record in Sage Intacct.'"
+                            [options]="reimbursableEmployeeOptions"
+                            [iconPath]="'people'"
+                            [placeholder]="'Select represntation'"
+                            [formControllerName]="'reimbursableEmployeeType'">
+                        </app-configuration-select-field>
+                        <div *ngIf="exportSettingForm.value.reimbursableEmployeeType === reimbursableEmployeeOptions[0].value">
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html:64]: same as above

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 4]
```diff
@@ -12,15 +12,23 @@ import { Sage300ExportSettingService } from 'src/app/core/services/sage300/sage3
...
 
   isOnboarding: boolean;
 
-  exportSettings: sage300ExportSettingGet;
+  exportSettings: any;
 
   exportSettingForm: FormGroup;
 
-  expenseGroupBy: Sage300ExportSettingFormOption[] = [
+  redirectLink: string = '';
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:23]: add enum

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 5]
```diff
@@ -12,15 +12,23 @@ import { Sage300ExportSettingService } from 'src/app/core/services/sage300/sage3
...
+
+  appName: string = AppName.SAGE300;
+
+  saveInProgress: boolean;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:27]: `isSaveInProgress`

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 6]
```diff
@@ -12,15 +12,23 @@ import { Sage300ExportSettingService } from 'src/app/core/services/sage300/sage3
 })
 export class Sage300ExportSettingsComponent implements OnInit {
 
-  isLoading: boolean;
+  isLoading: boolean = false;
 
   isOnboarding: boolean;
 
-  exportSettings: sage300ExportSettingGet;
+  exportSettings: any;
```
[the remark this answers is not in the document]
Comment (DhaaraniCIT) [reply src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:19]: this for now, once all the set will remove it

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 7]
```diff
@@ -33,102 +46,37 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+      this.exportSettings.credit_card_expense_grouped_by = ['expense_id'];
+      this.exportSettings.reimbursable_expense_grouped_by = ['expense_id'];
       this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup(this.exportSettings);
-      this.setCustomValidatorsAndWatchers();
+      this.exportSettingForm.controls.reimbursableExpense.setValidators(this.helper.exportSelectionValidator(this.exportSettingForm));
+      this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);
       this.isLoading = false;
     }, (error) => {
       this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup();
       this.isLoading = false;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:72]: we should set validator in error block too

File: src/app/core/models/sage300/sage300-configuration/sage300-export-setting.model.ts [hunk 8]
```diff
@@ -1,16 +1,20 @@
...
+import { ExpenseState, CCCExpenseState, Sage300ExpenseDate, Sage300ExportType, FyleField, ExpenseGroupingFieldOption } from "../../enum/enum.model";
+
 export type Sage300ExportSettingFormOption = {
     label: string,
-    value: ExpenseState | CCCExpenseState | Sage300ExportType | Sage300ExpenseDate | ExpenseGroupedBy | FyleField
+    value: ExpenseState | CCCExpenseState | Sage300ExportType | Sage300ExpenseDate | ExpenseGroupingFieldOption | FyleField
 }
 
+export type ValidatorRule = {
+  [expenseType: string]: string[];
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-export-setting.model.ts:10]: reimbursableExpense: string[];
creditCardExpense: string[];

File: src/app/core/models/sage300/sage300-configuration/sage300-export-setting.model.ts [hunk 9]
```diff
@@ -1,16 +1,20 @@
...
+import { FormControl, FormGroup } from "@angular/forms";
+import { ExpenseState, CCCExpenseState, Sage300ExpenseDate, Sage300ExportType, FyleField, ExpenseGroupingFieldOption } from "../../enum/enum.model";
+
 export type Sage300ExportSettingFormOption = {
     label: string,
-    value: ExpenseState | CCCExpenseState | Sage300ExportType | Sage300ExpenseDate | ExpenseGroupedBy | FyleField
+    value: ExpenseState | CCCExpenseState | Sage300ExportType | Sage300ExpenseDate | ExpenseGroupingFieldOption | FyleField
 }
 
+export type ValidatorRule = {
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-export-setting.model.ts:9]: ExportSettingValidatorRule

File: src/app/core/services/common/helper.service.ts [hunk 10]
```diff
@@ -45,4 +46,50 @@ export class HelperService {
     form.controls[controllerName].clearValidators();
     form.controls[controllerName].setValue(null);
   }
+
+  setCustomValidatorsAndWatchers(validatorRule: ValidatorRule, form: FormGroup) {
```
Comment (ashwin1111) [inline src/app/core/services/common/helper.service.ts:50]: `setExportSettingValidators`

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 11]
```diff
@@ -33,102 +46,37 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-            this.helper.markControllerAsRequired(this.exportSettingForm, element);
-          });
-        } else {
-          value.false.forEach((element: string) => {
-            this.helper.clearValidatorAndResetValue(this.exportSettingForm, element);
-          });
-        }
-      });
-    });
+  getExportType(exportType: string | null): string {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:51]: move it to helper service

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 12]
```diff
@@ -33,102 +46,37 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-      true: ['reimbursableExportType', 'reimbursableExportGroup', 'reimbursableExportDate', 'reimbursableExpenseState', 'defaultReimbursableCCCAccountName', 'defaultReimbursableCCCAccountId'],
-      false: ['reimbursableExportType', 'reimbursableExportGroup', 'reimbursableExportDate', 'reimbursableExpenseState', 'defaultReimbursableCCCAccountName', 'defaultReimbursableCCCAccountId']
-    },
-    'creditCardExpense': {
-      formController: 'creditCardExpense',
-      true: ['cccExportType', 'cccExportGroup', 'cccExportDate', 'cccExpenseState', 'defaultCreditCardCCCAccountName', 'defaultCreditCardCCCAccountId', 'defaultVendorName', 'defaultVendorId'],
-      false: ['cccExportType', 'cccExportGroup', 'cccExportDate', 'cccExpenseState', 'defaultCreditCardCCCAccountName', 'defaultCreditCardCCCAccountId', 'defaultVendorName', 'defaultVendorId']
-    }
-  };
+    private helper: HelperService  ) { }
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:49]: make it public, for simple methods like getExportType, can call it directly from html

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 13]
```diff
@@ -33,102 +46,37 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-  //     }
-  //   });
-  // }
-
-  // SetCustomValidatorsAndWatchers() {
-  //   // Toggles
-  //   This.createReimbursableExpenseWatcher();
-  //   This.createCreditCardExpenseWatcher();
-  // }
+  validatorRule: ValidatorRule = {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:55]: this need not be a global variable, add it as a const in setupPage()

**Comment by DhaaraniCIT** [comment]
all these comments are in export-setting-save PR [https://github.com/fylein/fyle-integrations-settings-app/pull/233/commits/a5bd3992ec40ca0f145151bb965386d5f82ac7f9](url)

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**